oauthenticator/generic.py

```diff
@@ -216,6 +224,46 @@ async def authenticate(self, handler, data=None):
 
         return user_info
 
+    # Refresh user access and refresh tokens (called periodically)
+    async def refresh_user(self, user, handler=None):
+        # Retrieve user authentication info and check if refresh is needed
+        auth_state = await user.get_auth_state()
+        expires_at = auth_state.get('expires_at', None)
+        refresh_token = auth_state.get('refresh_token', None)
+
+        # If no refresh_token, return success
+        if not refresh_token:
+            return True
+
+        # If no expiration, return success
+        if not expires_at:
+            return True
+
+        # If over 2x auth_refresh_age intervals from expiration, return success
+        if (time.time() + 2 * self.auth_refresh_age) < expires_at:
+            return True
+
+        self.log.info('Refreshing tokens for user %s', user.name)
+
+        # Attempt renewal
+        params = dict(
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            refresh_token=refresh_token,
+            grant_type='refresh_token',
+        )
+
+        headers = self._get_headers()
```

Comment:
this one is wrong. It has to be 
`headers = { "Content-Type": "application/x-www-form-urlencoded" }`